### PR TITLE
Clear leak of (copy of) OPENSSL_MALLOC_FAILURES env var string

### PR DESCRIPTION
**Pull request by sftcd**
I was playing with using the `OPENSSL_MALLOC_FAILURES` environment variable and noticed that the value of that string is leaked if it's set. This PR just provides a way to clear that and adds a call to do so to `s_client.c` which is what I was using for that test. That leak was a bit annoying as it meant all tests had a leak, whereas with this fix, any leaks resulting from memory allocation failures are more interesting, so valgrind output is more useful.

It's likely there's some better way trigger freeing that variable, so this is probably not quite the right thing to merge but if there's a good generic way to free stuff on exit then this might turn into a useful thing.

If (as is more likely) there's a better way to handle this, it's fine to just close this and fix that other way.

### CODE REVIEW

File: crypto/mem.c
Comment (tom-cosgrove-arm) [inline crypto/mem.c:170]: Please use `OPENSSL_strdup()` - not all systems have `strdup()`

Comment (sftcd) [reply crypto/mem.c]: sure, easily done (in push below)

File: crypto/mem.c
Comment (tom-cosgrove-arm) [inline crypto/mem.c:184]: Since `md_failstring` points into `md_fs_orig`, that should be set to `NULL` here too

File: crypto/mem.c
Comment (tom-cosgrove-arm) [inline crypto/mem.c:186]: Unnecessary `return`

**Comment by sftcd** [comment]
The push above makes the changes suggested today and adds a bit of documentation.

In the process of doing that it occurred to me that we could maybe do better if we just use a fixed sized buffer for `md_failstring`, rather than doing an `OPENSSL_strdup()` at all. That'd also mean no new API etc. So would it be acceptable to e.g. change to `static char md_failstring[MD_FAILSTRING_MAX]` with `MD_FAILSTRING_MAX` set to 128 or so and a check on the length of the output from `getenv()`?

**Comment by tom-cosgrove-arm** [comment]
@t8m would you have any thoughts [on this suggestion](https://github.com/openssl/openssl/pull/26039#issuecomment-2498249711)? I am nervous about adding to the static overhead of the library for a rarely-used feature, and I don't know what sort of length of malloc failure string people tend to set (i.e. 128 is almost certainly long enough all the time, but is quite a lot of memory - in some contexts - for an unused feature, but might not be enough for some tricky debugging situations - i.e. gradually extending the failure string to nearly track down an issue but then running out of space would be a very frustrating experience)

File: crypto/mem.c
Comment (t8m) [inline crypto/mem.c:170]: Please revert this to strdup(). The strdup() function was used here before already. Also calling any OPENSSL_ allocation function here makes it impossible to change the memory callbacks.

File: crypto/mem.c
Comment (t8m) [reply crypto/mem.c:170]: > Please use `OPENSSL_strdup()` - not all systems have `strdup()`

We cannot use OPENSSL_strdup() here.

File: crypto/mem.c
Comment (tom-cosgrove-arm) [reply crypto/mem.c]: Sorry, that was my fault - I asked for that change

**Comment by tom-cosgrove-arm** [comment]
> would you prefer I close this PR and open another or force-push that change to this PR?

No need to open a new PR, just update this one

**Comment by sftcd** [comment]
did a force push with the static version (will fix any CI issues in a sec)

**Comment by sftcd** [comment]
OK, the CI seems happy now, so this static-flavoured PR is ready to review...

**Review by t8m — DISMISSED** [review]

File: crypto/mem.c
Comment (tom-cosgrove-arm) [inline crypto/mem.c:42]: (minor) Why not initialise this to `NULL` (or leave it uninitialised, like it was before)? Then we keep the behaviour that if this is `NULL` there is no `OPENSSL_MALLOC_FAILURES`

File: crypto/mem.c
Comment (tom-cosgrove-arm) [inline crypto/mem.c:173]: (minor) It's probably safer (in terms of future changes to this code) to use `sizeof(md_failbuf) - 1` on both this line and the one below

**Review by tom-cosgrove-arm — APPROVED** [review]
A couple of minor comments, but happy either way

File: crypto/mem.c
Comment (sftcd) [reply crypto/mem.c:42]: ack, will do

File: crypto/mem.c
Comment (sftcd) [reply crypto/mem.c:173]: will do

**Comment by sftcd** [comment]
fixup above is for @tom-cosgrove-arm minor comments, sorry if that forces you to re-approve (and/or if I mucked up and have to push again to make CI happy:-)

I also have a question though: I have a script that iterates causing one more malloc to fail each time (that's [in a gist](https://gist.github.com/sftcd/5d951527ad76c1f9434fce75b2e508bf)). When I run that valgrind repeatedly reports some errors that are a nuisance, (I'm using the script to check my ECH client-side code), likely not serious but probably worth fixing. I have some fixes for those that I think get me a clean run of the script (running now again), but I'm not very sure that these are the right fixes. You can see the relevant changes [here](https://github.com/openssl/openssl/commit/6d9134e8d7fd86469c513a16231ad1b1e3e02e8f). I guess the right thing here is to let you re-approve this, then open a separate PR for those fixes after this one is merged?

But if you prefer to add those changes to this PR, that's fine by me too.

**Comment by tom-cosgrove-arm** [comment]
> then open a separate PR for those fixes after this one is merged?

Yes, best open separate PRs to fix different issues

**Review by shahsb — APPROVED** [review]
LGTM.

File: crypto/mem.c
Comment (tom-cosgrove-arm) [reply crypto/mem.c:173]: > Got some windows CI issue with that change so reverted it

What was the issue? I can't think why it would cause a failure (I didn't get to see this before the next run hid the logs, and I don't know how to find the CI logs for an old version of a PR)

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**
